From 2a3658c003ad8cfc3251033462499727620ecd42 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Tue, 9 Dec 2025 20:28:52 +0800
Subject: [PATCH 096/105] wifi: mt76: mt7996: check ddone for tx queues

Check ddone for tx queues.
No issue encountered, just align logan configuraion to prevent
potential issue.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 dma.c        | 4 ++++
 mt76.h       | 1 +
 mt7996/dma.c | 8 ++++----
 3 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/dma.c b/dma.c
index dbbafad1c..2659e3f62 100644
--- a/dma.c
+++ b/dma.c
@@ -422,6 +422,10 @@ mt76_dma_tx_cleanup(struct mt76_dev *dev, struct mt76_queue *q, bool flush)
 		last = Q_READ(q, dma_idx, MT_QUEUE_DMA_IDX);
 
 	while (q->queued > 0 && q->tail != last) {
+		if ((q->flags & MT_QFLAG_CHECK_DDONE) &&
+		    !(q->desc[q->tail].ctrl & cpu_to_le32(MT_DMA_CTL_DMA_DONE)))
+			break;
+
 		mt76_dma_tx_cleanup_idx(dev, q, q->tail, &entry);
 		mt76_npu_txdesc_cleanup(q, q->tail);
 		mt76_queue_tx_complete(dev, q, &entry);
diff --git a/mt76.h b/mt76.h
index de65719eb..49bd41ed1 100644
--- a/mt76.h
+++ b/mt76.h
@@ -72,6 +72,7 @@
 #define MT_QFLAG_WED_RRO_EN	BIT(7)
 #define MT_QFLAG_EMI_EN		BIT(8)
 #define MT_QFLAG_NPU		BIT(9)
+#define MT_QFLAG_CHECK_DDONE	BIT(10)
 
 #define __MT_WED_Q(_type, _n)	(MT_QFLAG_WED | \
 				 FIELD_PREP(MT_QFLAG_WED_TYPE, _type) | \
diff --git a/mt7996/dma.c b/mt7996/dma.c
index c21e1b28e..5862e873a 100644
--- a/mt7996/dma.c
+++ b/mt7996/dma.c
@@ -11,7 +11,7 @@ int mt7996_init_tx_queues(struct mt7996_phy *phy, int idx, int n_desc,
 			  int ring_base, struct mtk_wed_device *wed)
 {
 	struct mt7996_dev *dev = phy->dev;
-	u32 flags = 0;
+	u32 flags = MT_QFLAG_CHECK_DDONE;
 	int i;
 
 	if (phy->mt76->band_idx == MT_BAND1 && !dev->hif2 && is_mt7996(&dev->mt76)) {
@@ -26,13 +26,13 @@ int mt7996_init_tx_queues(struct mt7996_phy *phy, int idx, int n_desc,
 		idx -= MT_TXQ_ID(0);
 
 		if (wed == &dev->mt76.mmio.wed_hif2)
-			flags = MT_WED_Q_TX(0);
+			flags |= MT_WED_Q_TX(0);
 		else
-			flags = MT_WED_Q_TX(idx);
+			flags |= MT_WED_Q_TX(idx);
 	}
 
 	if (mt76_npu_device_active(&dev->mt76))
-		flags = MT_NPU_Q_TX(phy->mt76->band_idx);
+		flags |= MT_NPU_Q_TX(phy->mt76->band_idx);
 
 	return mt76_connac_init_tx_queues(phy->mt76, idx, n_desc,
 					  ring_base, wed, flags);
-- 
2.45.2

